dashboard/modules/job/job_manager.py

```diff
@@ -375,12 +371,25 @@ async def run(
             )
 
             if self._stop_event.is_set():
-                polling_task.cancel()
                 if sys.platform == "win32" and self._win32_job_object:
+                    polling_task.cancel()
                     win32job.TerminateJobObject(self._win32_job_object, -1)
                 elif sys.platform != "win32":
-                    # TODO (jiaodong): Improve this with SIGTERM then SIGKILL
-                    child_process.kill()
+                    try:
+                        os.killpg(os.getpgid(child_process.pid), signal.SIGTERM)
+                    except ProcessLookupError:
+                        # Process already completed.
+                        pass
+                    else:
+                        # Wait for job to terminate, otherwise kill process forcefully
+                        # after timeout.
+                        try:
+                            await asyncio.wait_for(
+                                polling_task, self.WAIT_FOR_JOB_TERMINATION_S
+                            )
+                        except TimeoutError:
```

Comment:
add a log message here that the timeout is up and it's being force-killed

Reply:
also add one in the case that the process _did_ terminate gracefully. this may help users who are trying to figure out what happened, getting teardown / signal handling right can be very painful (as you learned when trying to figure out where your `KeyboardInterrupt` was going...

Reply:
Done!